### PR TITLE
Fix: Support insert overwrite with dynamic partitions in the BigQuery adapter

### PR DESCRIPTION
**Pull request by izeigerman**

### CODE REVIEW

File: sqlmesh/core/engine_adapter/base.py [hunk 1]
```diff
@@ -976,7 +977,7 @@ def _get_temp_table(
         """
         Returns the name of the temp table that should be used for the given table name.
         """
-        table = t.cast(exp.Table, exp.to_table(table).copy())
+        table = t.cast(exp.Table, exp.to_table(table, dialect=self.dialect).copy())
```
Comment (tobymao) [inline sqlmesh/core/engine_adapter/base.py:980]: don't add dialect here, everything here is already normalized and agnostic

File: sqlmesh/core/snapshot/evaluator.py [hunk 2]
```diff
@@ -800,8 +800,8 @@ def insert(
         **kwargs: t.Any,
     ) -> None:
         if isinstance(model.kind, IncrementalUnmanagedKind) and model.kind.insert_overwrite:
-            self.adapter.insert_overwrite(
-                name, query_or_df, columns_to_types=model.columns_to_types
+            self.adapter.insert_overwrite_by_partition(
+                name, query_or_df, model.partitioned_by, columns_to_types=model.columns_to_types
```
Comment (eakmanrq) [inline sqlmesh/core/snapshot/evaluator.py:804]: I'm sure this has been asked before, but just confirming that a model in dbt must have a partition_by if it is incremental?

File: tests/core/engine_adapter/test_bigquery.py [hunk 3]
```diff
@@ -399,3 +401,17 @@ def test_begin_end_session(mocker: MockerFixture):
     execute_b_call = connection_mock._client.query.call_args_list[2]
     assert execute_b_call[1]["query"] == "SELECT 3;"
     assert not execute_b_call[1]["job_config"].connection_properties
+
+
+def _to_sql_calls(execute_mock: t.Any, identify: bool = True) -> t.List[str]:
```
Comment (eakmanrq) [inline tests/core/engine_adapter/test_bigquery.py:406]: Thanks for cleaning this up.

File: sqlmesh/core/snapshot/evaluator.py (hunk 2 above)
Comment (izeigerman) [reply sqlmesh/core/snapshot/evaluator.py:804]: We validate this here: https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/model/definition.py#L626-L634

**Review by tobymao — COMMENTED** [review]
""

File: sqlmesh/core/engine_adapter/bigquery.py [hunk 4]
```diff
@@ -354,6 +354,44 @@ def merge(
...
+            if columns_to_types is None:
+                columns_to_types = self.columns(temp_table_name)
+
+            partition_type_sql = columns_to_types[partition_column.name].sql(dialect=self.dialect)
+            temp_table_name_sql = temp_table_name.sql(dialect=self.dialect)
+            self.execute(
+                f"DECLARE target_partitions ARRAY<{partition_type_sql}> DEFAULT (SELECT AS STRUCT ARRAY_AGG(DISTINCT {partition_sql}) FROM {temp_table_name_sql});"
+            )
+
+            where = exp.In(
```
Comment (tobymao) [inline sqlmesh/core/engine_adapter/bigquery.py:383]: the ast here is wrong for in unnest, i'll add a helper for this or you can do

exp.condition(...)

Comment (izeigerman) [reply sqlmesh/core/engine_adapter/bigquery.py:383]: why is it wrong? It results in a correct sql

Comment (izeigerman) [reply sqlmesh/core/engine_adapter/bigquery.py:383]: Not sure how to apply `exp.condition` in this case

Comment (tobymao) [reply sqlmesh/core/engine_adapter/bigquery.py:383]: in general building ast is dangerous and error prone

Comment (tobymao) [reply sqlmesh/core/engine_adapter/bigquery.py:383]: v17.6.0

partition_exp.isin(unnest="target_partitions")

File: sqlmesh/core/engine_adapter/bigquery.py [hunk 5]
```diff
@@ -354,6 +354,44 @@ def merge(
...
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None,
+    ) -> None:
+        if len(partitioned_by) != 1:
+            raise SQLMeshError(
+                f"Bigquery only supports partitioning by one column, {len(partitioned_by)} were provided."
+            )
+
+        partition_exp = partitioned_by[0]
+        partition_sql = partition_exp.sql(dialect=self.dialect)
+        partition_column = list(partition_exp.find_all(exp.Column))[0]
```
Comment (tobymao) [inline sqlmesh/core/engine_adapter/bigquery.py:371]: partition_column = partition_exp.find(exp.Column)

**Review by tobymao — APPROVED** [review]

**Pull request merged**